packages/Chem/src/rdkit-service/rdkit-service-worker-substructure.ts

```diff
@@ -29,95 +28,116 @@ export class RdKitServiceWorkerSubstructure extends RdKitServiceWorkerSimilarity
+            mol.delete();
+          } else
+            this._rdKitMols![i] = mol;
+        } else {
+          if (useSubstructLib)
+            this._malformedIdxs!.setFast(i, true);
+          numMalformed++;
+        }
       }
     }
-    return malformed;
+    return numMalformed;
   }
 
-  searchSubstructure(queryMolString: string, queryMolBlockFailover: string, bitset?: boolean[]): string {
-    const matches: number[] = [];
-    if (this._rdKitMols === null)
-      return '[' + matches.join(', ') + ']';
+  searchSubstructure(queryMolString: string, queryMolBlockFailover: string, molecules?: string[],
+    useSubstructLib?: boolean): Uint32Array {
+    if (this._rdKitMols === null && !useSubstructLib && !molecules)
+      throw new Error('Chem | Molecules for substructure serach haven\'t been provided');
 
-    let queryMol = null;
+    let queryMol = getQueryMolSafe(queryMolString, queryMolBlockFailover, this._rdKitModule);
+    
+    if (queryMol !== null) {
+      const matches = useSubstructLib ? this.searchWithSubstructLib(queryMol) :
+        molecules ? this.searchWithPatternFps(queryMol, molecules) : this.searchWithoutPatternFps(queryMol);
+      queryMol.delete();
+      return matches;
+    } else
+      throw new Error('Chem | Search pattern cannot be set');
+  }
 
-    if (isMolBlock(queryMolString)) {
-      if (queryMolString.includes(' H ') || queryMolString.includes('V3000'))
-        queryMol = getMolSafe(queryMolString, {mergeQueryHs: true}, this._rdKitModule).mol;
-      else {
-        try {
-          queryMol = this._rdKitModule.get_qmol(queryMolString);
-          queryMol.convert_to_aromatic_form();
-        } catch (e) {
-          if (queryMol) {
-            queryMol.delete();
-            queryMol = null;
-          }
-        }
-      }
-    } else { // not a molblock
-      try {
-        queryMol = this._rdKitModule.get_qmol(queryMolString);
-      } catch (e) {}
-      if (queryMol !== null) {
-        const mol = getMolSafe(queryMolString, {mergeQueryHs: true}, this._rdKitModule).mol;
-        if (mol !== null) { // check the qmol is proper
-          const match = mol.get_substruct_match(queryMol);
-          if (match === '{}') {
-            queryMol.delete(); //remove mol object previously stored in queryMol
-            queryMol = mol;
-          } else
-            mol.delete();
-        } // else, this looks to be a real SMARTS
-      } else { // failover to queryMolBlockFailover
-        // possibly get rid of fall-over in future
-        queryMol = getMolSafe(queryMolBlockFailover, {mergeQueryHs: true}, this._rdKitModule).mol;
+  searchWithSubstructLib(queryMol: RDMol): Uint32Array {
+    const matches = new BitArray(this._substructLibrary!.size() + this._malformedIdxs!.trueCount());
+    const str = this._substructLibrary!.get_matches(queryMol, undefined, undefined, this._substructLibrary!.size());
+    const matchesIdxs = JSON.parse(str);
```

Comment:
Why are we passing indexes in JSON?

Reply:
Here the SubstructLibrary itself returns result in form of array of indexes as a string (this is how get_matches is implemented in rdkit minimal lib). The only way for us to get these indexes is parse a JSON string. And after that we convert indexes to BitArray.